### ISSUE TITLE
[WebGPU] Unify Emscripten and dawn-native paths in the runtime

### ISSUE THREAD

**Issue by jrprice** [halide/Halide#7248]
There are a few differences in the native APIs implemented by Emscripten and Dawn that mean Halide's WebGPU backend currently requires a build-time switch to select between the two. Once these differences are resolved, we should be able to remove this conditional logic and the build-time switch.

The differences are:
- Emscripten missing `wgpuCreateInstance` and `wgpuReleaseInstance`
- Emscripten missing `wgpuAdapterGetLimits`
- Waiting for async device init callbacks: `emscripten_sleep` vs `usleep`
- Waiting on other async callbacks: `emscripten_sleep()` vs `wgpuDeviceTick`

**Comment by steven-johnson** [comment]
Is there any plan as to when/if this will get resolved?

**Comment by steven-johnson** [comment]
One comment in https://github.com/halide/Halide/pull/6492 suggests that the lack of support for `wgpuInstanceProcessEvents()` is also tied to this issue, but I don't see it mentioned above

**Comment by jrprice** [comment]
> lack of support for wgpuInstanceProcessEvents() is also tied to this issue

Yeah, this is the expected name for the solution for the `emscripten_sleep()` vs `wgpuDeviceTick()` discrepancy. I've added it to the original comment to make that clearer.

> Is there any plan as to when/if this will get resolved?

We're focusing on getting V1 of WebGPU shipped in Chrome really soon [1], and then we expect to have a renewed focus on making the native vs web story better. So yes, these things will get resolved, but I can't give an exact timeframe right now.

[1]: [Intent to ship: WebGPU (Chromium)](https://groups.google.com/a/chromium.org/g/blink-dev/c/VomzPhvJCxI/m/SUhU9Z0vAgAJ)

**Comment by steven-johnson** [comment]
Ah, gotcha, thanks. FYI, googling for `wgpuInstanceProcessEvents()` shows that it is in Emscripten's current version of `webgpu.h`: https://github.com/emscripten-core/emscripten/blob/main/system/include/webgpu/webgpu.h -- is it already implemented there?

**Comment by jrprice** [comment]
> is it already implemented there?

It's in the header (which I think is just a copy of Dawn's header), but the implementation is still absent:
https://github.com/emscripten-core/emscripten/blob/7de3bc93f0f7b65549215772255779ad8305ce78/src/library_webgpu.js#L2426

**Comment by steven-johnson** [comment]
As of Aug 2023, Emscripten/Dawn now support (and *require*) wgpuCreateInstance() and wgpuInstanceRelease() in non-native code, so I updated those in [#????](https://github.com/halide/Halide/pull/7804). (wgpuInstanceProcessEvents() and wgpuAdapterGetLimits() are still unimplemented.)

**Comment by steven-johnson** [comment]
As of now,  wgpuInstanceProcessEvents() and wgpuAdapterGetLimits() are both present in webgpu.h -- are they still nonfunctional?

**Comment by jrprice** [comment]
`wgpuInstanceProcessEvents()` and `wgpuAdapterGetLimits()` are still both unimplemented by Emscripten.

For future reference the Emscripten implementation is [here](https://github.com/emscripten-core/emscripten/blob/main/src/library_webgpu.js), and it's usually easy to search for these functions since they are present but both just `abort()`.

**Comment by steven-johnson** [comment]
Since WebGPU API issues have come up again, I see these are still unimplemented; is there any likely path to them getting there, or should we assume this is just a long-term API divergence?

**Comment by jrprice** [comment]
I've asked the folks closer to this, and the response is still that we should eventually get to a world where you can target both native and Emscripten without any `#ifdef`s. I'm afraid we can't give a specific timeframe for when this will happen though. The latest header update PR did remove one such `#ifdef` however, so that's something!
